src/libraries/icubmod/bcbBattery/bcbBattery.cpp

```diff
@@ -65,214 +80,111 @@ bool BcbBattery::open(yarp::os::Searchable& config)
-
-        //empty the buffer
-        char c = 0;
-        int r = 0;
-        do
-        {
-            r = pSerial->receiveChar(c);
-        } while (r != 0);
+        yInfo("BcbBattery starting transmission");
+        startTransmission();
+        yInfo("BcbBattery started successfully");
     }
     else
     {
+#ifndef DEBUG_TEST
         yError("BcbBattery pSerial == NULL");
         return false;
+#endif
     }
 
-
     return true;
 }
 
-void BcbBattery::run()
+void batteryReaderThread::run()
 {
     double timeNow=yarp::os::Time::now();
-    lock_guard<mutex> lck(mtx);
-
-    //if 100ms have passed since the last received message
-    if (timeStamp+0.1<timeNow)
-    {
-        //status=IBattery::BATTERY_TIMEOUT;
-    }
+    int recb = 0;
 
     //read battery data.
-    serial_buff[0] = 0;
-
-    int output[10] = {0};
-    bool syncError = false;
-
+#ifndef DEBUG_TEST
     if (pSerial)
     {
-        size_t i = 0;
-        bool sync0 = false;
-        bool syncr = false;
-        bool syncn = false;
-        do
-        {
-            output[0] = pSerial->receiveChar(serial_buff[0]);
-            sync0 = output[0] && (serial_buff[0] == '\0');
-            ++i;
-        } while (!(sync0) && (i < 20) && !isClosing);
-
-        if (isClosing)
-        {
-            return;
-        }
-
-        if (sync0)
-        {
-            output[1] = pSerial->receiveChar(serial_buff[1]); //voltage
-            output[2] = pSerial->receiveChar(serial_buff[2]); //voltage
-            output[3] = pSerial->receiveChar(serial_buff[3]); //current
-            output[4] = pSerial->receiveChar(serial_buff[4]); //current
-            output[5] = pSerial->receiveChar(serial_buff[5]); //charge
-            output[6] = pSerial->receiveChar(serial_buff[6]); //charge
-            output[7] = pSerial->receiveChar(serial_buff[7]); //status
-        }
-
-        i = 0;
-        do
-        {
-            output[8] = pSerial->receiveChar(serial_buff[8]);
-            syncr = output[8] && (serial_buff[8] == '\r');
-            if (syncr)
-            {
-                output[9] = pSerial->receiveChar(serial_buff[9]);
-                syncn = output[9] && (serial_buff[9] == '\n');
-
-                if (!syncn)
-                {
-                    if (!syncError && !silenceSyncWarnings)
-                    {
-                        yDebug() << "Sync error (n)";
-                    }
-                    syncError = true;
-                }
-            }
-            else
-            {
-                if (!syncError && !silenceSyncWarnings)
-                {
-                    yDebug() << "Sync error (r)";
-                }
-                syncError = true;
-            }
-            ++i;
-        } while (!(syncn) && (i < 20) && !isClosing);
-
-        if (isClosing)
-        {
-            return;
-        }
-
-        serial_buff[10] = 0;
+        recb = pSerial->receiveBytes(tmp_buff, buff_len);
```

Comment:
Just to understand, for how long this would be blocking?

Reply:
For the time specified in the configuration file by `readtimeoutmsec` parameters. e.g. 5ms.
The method will extract all the characters from the buffer.

Reply:
Clear, thanks!